### ISSUE TITLE
Issue with padding ?

### ISSUE THREAD

**Issue by tlepley-cadence** [pytorch/glow#2037]
I've got an ONNX model with standalone Pad operations. I'm currently adding the support in Glow, including an optimization to merge this standalone Pad with Convolution. 

Reading at ONNX and looking at the Glow code, I think that the Glow convolution doesn't correctly support ONNX padding (I didn't look yet at the Caffe2 padding). 

In particular, the Glow code assumes that padding in the model is just given for spatial dimensions (array of 4 numbers), while ONNX specifies that the padding should be given for all dimensions (array of 8 values for a 4D tensor):
> The value represent the number of pixels added to the beginning and end part of the corresponding axis. `pads` format should be as follow [x1_begin, x2_begin...x1_end, x2_end,...], where xi_begin the number of pixels added at the beginning of axis `i` and xi_end, the number of pixels added at the end of axis `i`.

I would like to make sure that we all agree on the interpretation of the ONNX spec, even if we decide for now that we just want to support spatial padding in Glow. For instance, if we want to grow the tensor spatially, the 'pads' attribute in the model would be (considering NCHW) :
[ 0, 0, 1, 1, 0, 0, 1, 1] , and not [1, 1, 1, 1]

Any opinion ?

**Comment by tlepley-cadence** [comment]
I just looked at different models:
- ONNX models from Glow
- ONNX model generated my ft2onnx (one of the converters from the pytorch project)

Actually, while the description of the padding attribute in the ONNX documentation looks similar for the various operators, I get a padding vector of a different size depending on the situation for 4D tensors. There is IMO a problem and I think that we should always get a 8 element vector for 4D tensors.

**Glow zfnet512_onnx_model.txt:** 
- _Convolution_ op:  attribute = name: "pads" ints: 1 ints: 1 ints: 1 ints: 1 type: INTS 

**my_converted_onnx_model.txt:** 
- _Pad_ op: attribute = name: "pads" ints: 0 ints: 0 ints: 1 ints: 1 ints: 0 ints: 0 ints: 1 ints: 1 type: INTS 
- _Maxpool_ op: attribute = name: "pads" ints: 0 ints: 0 ints: 0 ints: 0 type: INTS

**Comment by nadavrot** [comment]
CC @beicy

**Comment by beicy** [comment]
@tlepley-cadence Thanks for posting this issue.

 So far, Glow only supports 2D convolutions.  For 2D convolution, the pads is for W*H, which is 4 numbers (the pads is based on the kernel dims). I think ONNX spec also means that if we take a look at their example: https://github.com/onnx/onnx/blob/master/docs/Operators.md#Conv
In addition,  the 2D conv ops in the ONNX models we are working on are with 4-number pads and I haven't seen any extra flag to indicate the pads should have 8 numbers.  Therefore, I think our current implantation works. Of course, later after the 3D convolution is supported, the corresponding pads will be improved as well. 

In terms of the model you mentioned: 
```
my_converted_onnx_model.txt:

Pad op: attribute = name: "pads" ints: 0 ints: 0 ints: 1 ints: 1 ints: 0 ints: 0 ints: 1 ints: 1 type: INTS
Maxpool op: attribute = name: "pads" ints: 0 ints: 0 ints: 0 ints: 0 type: INTS
```
I think the  `Pad `op is different from the "pads" arg used in Conv op, it should be the following one:  
https://github.com/onnx/onnx/blob/master/docs/Operators.md#Pad

If you like, would you please provide a link for the converter ("`ONNX model generated my ft2onnx (one of the converters from the pytorch project)`" ) and the converted model? I can double check it :)

**Comment by tlepley-cadence** [comment]
For the converter, I used TF2ONNX : https://github.com/onnx/tensorflow-onnx
For the model, I can't share.

IMO, the main issue is that the ONNX specification doesn't explicitly tell that only the padding for spatial dimensions should be specified for operations like convolution. I then think that this is more a question for ONNX than for Glow. I'm going to file an issue to ONNX.

For now, it seems that the converter I'm using and Glow agree on the interpretation of ONNX, so I guess there is no actual issue :-) I'll then close this issue for now.
